plugins/pluto/src/ad936x.cpp

```diff
@@ -0,0 +1,596 @@
+	controlWidgetLayout->setMargin(0);
+	controlWidgetLayout->setContentsMargins(0, 0, 0, 0);
+	controlsWidget->setLayout(controlWidgetLayout);
+
+	QScrollArea *scrollArea = new QScrollArea(this);
+	scrollArea->setWidgetResizable(true);
+	scrollArea->setWidget(controlsWidget);
+
+	controlsLayout->addWidget(scrollArea);
+
+	///  fist widget the global settings can be created with iiowigets only
+	controlWidgetLayout->addWidget(generateGlobalSettingsWidget(controlsWidget));
+
+	/// second is Rx ( receive chain)
+	controlWidgetLayout->addWidget(generateRxChainWidget(controlsWidget));
+
+	/// third is Tx (transimt chain)
+	controlWidgetLayout->addWidget(generateTxChainWidget(controlsWidget));
+
+	controlWidgetLayout->addItem(new QSpacerItem(1, 1, QSizePolicy::Preferred, QSizePolicy::Expanding));
+
+	m_blockDiagramWidget = new QWidget(this);
+	Style::setBackgroundColor(m_blockDiagramWidget, json::theme::background_primary);
+	QVBoxLayout *blockDiagramLayout = new QVBoxLayout(m_blockDiagramWidget);
+	m_blockDiagramWidget->setLayout(blockDiagramLayout);
+
+	QWidget *blockDiagramWidget = new QWidget(this);
+	QVBoxLayout *blockDiagramWidgetLayout = new QVBoxLayout(blockDiagramWidget);
+	blockDiagramWidget->setLayout(blockDiagramWidgetLayout);
+
+	QScrollArea *blockDiagramWidgetScrollArea = new QScrollArea(this);
+	blockDiagramWidgetScrollArea->setWidgetResizable(true);
+	blockDiagramWidgetScrollArea->setWidget(blockDiagramWidget);
+
+	blockDiagramLayout->addWidget(blockDiagramWidgetScrollArea);
+
+	QLabel *blockDiagram = new QLabel(m_blockDiagramWidget);
+	blockDiagramWidgetLayout->addWidget(blockDiagram);
+	blockDiagram->setAlignment(Qt::AlignCenter);
+	QPixmap pixmap(":/pluto/ad936x.svg");
+	blockDiagram->setPixmap(pixmap);
+
+	centralWidget->addWidget(m_controlsWidget);
+	centralWidget->addWidget(m_blockDiagramWidget);
+
+	m_tool->addWidgetToCentralContainerHelper(centralWidget);
+
+	QButtonGroup *centralWidgetButtons = new QButtonGroup(this);
+	centralWidgetButtons->setExclusive(true);
+
+	QPushButton *ad963xBtn = new QPushButton("Controls", this);
+	ad963xBtn->setCheckable(true);
+	ad963xBtn->setChecked(true);
+	Style::setStyle(ad963xBtn, style::properties::button::blueGrayButton);
+	connect(ad963xBtn, &QPushButton::clicked, this,
+		[=, this]() { centralWidget->setCurrentWidget(m_controlsWidget); });
+
+	QPushButton *blockDiagramBtn = new QPushButton("Block Diagram", this);
+	blockDiagramBtn->setCheckable(true);
+	Style::setStyle(blockDiagramBtn, style::properties::button::blueGrayButton);
+	connect(blockDiagramBtn, &QPushButton::clicked, this,
+		[=, this]() { centralWidget->setCurrentWidget(m_blockDiagramWidget); });
+
+	centralWidgetButtons->addButton(ad963xBtn);
+	centralWidgetButtons->addButton(blockDiagramBtn);
+
+	m_tool->addWidgetToTopContainerHelper(ad963xBtn, TTA_LEFT);
+	m_tool->addWidgetToTopContainerHelper(blockDiagramBtn, TTA_LEFT);
+}
+
+AD936X::~AD936X()
+{
+	// close Connection
+	ConnectionProvider::close(m_uri);
```

Comment:
I've seen this here (in the AD936X class), but also in all the other classes. ConnectionProvider::close needs to be called as many times as ConnectionProvider::open is called. Each "open" increases the context usage reference count. So we're going to have Scopy hanging if instances are not properly closed/dereferenced.
You either track all the "open" calls and call "close". The other option, that's been used across all plugins, is to handle the "open" in the Plugin class and pass down to all the other classes just the instance of iio_context.